querier_test.go

```diff
@@ -200,3 +203,641 @@ func expandSeriesIterator(it SeriesIterator) (r []sample, err error) {
 
 	return r, it.Err()
 }
+
+// Index: labels -> postings -> chunkMetas -> chunkRef
+// ChunkReader: ref -> vals
+func createIdxChkReaders(tc []struct {
+	lset   map[string]string
+	chunks [][]sample
+}) (IndexReader, ChunkReader) {
+	sort.Slice(tc, func(i, j int) bool {
+		return labels.Compare(labels.FromMap(tc[i].lset), labels.FromMap(tc[i].lset)) < 0
+	})
+
+	postings := &memPostings{m: make(map[term][]uint32, 512)}
+	chkReader := mockChunkReader(make(map[uint64]chunks.Chunk))
+	mi := newMockIndex()
+
+	for i, s := range tc {
+		metas := make([]*ChunkMeta, 0, len(s.chunks))
+		for _, chk := range s.chunks {
+			// Collisions can be there, but for tests, its fine.
+			ref := rand.Uint64()
+
+			metas = append(metas, &ChunkMeta{
+				MinTime: chk[0].t,
+				MaxTime: chk[len(chk)-1].t,
+				Ref:     ref,
+			})
+
+			chunk := chunks.NewXORChunk()
+			app, _ := chunk.Appender()
+			for _, smpl := range chk {
+				app.Append(smpl.t, smpl.v)
+			}
+			chkReader[ref] = chunk
+		}
+
+		mi.AddSeries(uint32(i), labels.FromMap(s.lset), metas...)
+
+		postings.add(uint32(i), term{})
+		for _, l := range labels.FromMap(s.lset) {
+			postings.add(uint32(i), term{l.Name, l.Value})
+		}
+	}
+
+	for tm := range postings.m {
+		mi.WritePostings(tm.name, tm.name, postings.get(tm))
+	}
+
+	return mi, chkReader
+}
+
+func TestBlockQuerier(t *testing.T) {
+	newSeries := func(l map[string]string, s []sample) Series {
+		return &mockSeries{
+			labels:   func() labels.Labels { return labels.FromMap(l) },
+			iterator: func() SeriesIterator { return newListSeriesIterator(s) },
+		}
+	}
+
+	// Build the querier on data first. Then execute queries on it.
+	basedata := [][]struct {
+		lset   map[string]string
+		chunks [][]sample
+	}{
+		{
+			{
+				lset: map[string]string{
+					"a": "a",
+				},
+				chunks: [][]sample{
+					{
+						{1, 2}, {2, 3}, {3, 4},
+					},
+					{
+						{5, 2}, {6, 3}, {7, 4},
+					},
+				},
+			},
+			{
+				lset: map[string]string{
+					"a": "a",
+					"b": "b",
+				},
+				chunks: [][]sample{
+					{
+						{1, 1}, {2, 2}, {3, 3},
+					},
+					{
+						{5, 3}, {6, 6},
+					},
+				},
+			},
+			{
+				lset: map[string]string{
+					"b": "b",
+				},
+				chunks: [][]sample{
+					{
+						{1, 3}, {2, 2}, {3, 6},
+					},
+					{
+						{5, 1}, {6, 7}, {7, 2},
+					},
+				},
+			},
+		},
+	}
+
+	cases := []struct {
+		dataIdx int
+
+		mint, maxt int64
+		ms         []labels.Matcher
+		exp        SeriesSet
+	}{
+		{
+			dataIdx: 0,
```

Comment:
No need to complicate things for multiple datasets if we are not using it yet. We never might.